### ISSUE TITLE
djxl PNG Conversion Acting Up

### ISSUE THREAD

**Issue by gianni-rosato** [libjxl/libjxl#2289]
**Describe the bug**
When converting a particular JXL image to PNG, the PNG has noticeable color issues with patches of discoloration.

**To Reproduce**
Download [this](https://cdn.discordapp.com/attachments/794206170445119489/1083872278128885790/imagetest15-Q70.jxl)

djxl theimage.jxl out.png

Compare the png to the JXL.

**Expected behavior**
The PNG would be visually identical to the JXL, or at least close enough where I wouldn't see obvious discoloration or where the PNG is earning horrible ssimu2 scores.

**Screenshots**
If applicable, add screenshots or example input/output images to help explain your problem.

**Environment**
 - OS: Linux
 - Compiler version: gcc (GCC) 12.2.1 20230201
 - CPU type: x86_64
 - cjxl/djxl version string: djxl v0.9.0 13caf6f8 [AVX2,SSE4,SSSE3,SSE2]

**Additional context**
The problem persists with other versions of cjxl/djxl when using the source PNG (which is almost 100mb).

**Comment by thebombzen** [comment]
Notably, the JXL file has an attached ICC profile and is XYB encoded. djxl is outputting a PNG in Linear Light rather than in sRGB.

**Comment by gianni-rosato** [comment]
If you want the source PNG, let me know

**Comment by novomesk** [comment]
It is the fact that djxl uses `RGB_D65_SRG_Rel_Lin` profile there.

If the JXL and decoded PNG look significantly visually different, that's a problem of non-implemented (or not working) color management support of the viewer application.

The functionality of djxl is not technically wrong.

Color profile used to decode can be changed:

`djxl --color_space=RGB_D65_SRG_Rel_SRG imagetest15-Q70.jxl out2.png`

**Comment by gianni-rosato** [comment]
When inputting the weird PNG into ssimulacra2, it doesn't understand it & gives it a terrible score

**Comment by kmilos** [comment]
Did you try the "keep original" color space option when exporting from darktable? The JPEG XL file will be considerably larger, but it'll preserve the profile. Otherwise, it discards it when encoding and only decodes to linear RGB (unless you're doing lossless, i.e. quality 100 as well when you get out what you put in).

**Comment by jonsneyers** [comment]
There are three things that would fix this, and I think we need to do all of them:
1. Make cjxl recognize Display P3 and use the Enum representation for it rather than a generic icc profile. This would help for compression, and it circumvents the djxl bug.
2. Make djxl default to writing a 16-bit png if the image is in xyb colorspace.
3. Make libjxl decode do full color management by default (make it a special build flag to build without dependency on external CMS).

**Comment by thebombzen** [comment]
> There are three things that would fix this, and I think we need to do all of them:
> 2. Make djxl default to writing a 16-bit png if the image is in xyb colorspace.

I would disagree with always decoding XYB to 16-bit PNGs. I think it should only do this if there's an attached ICC profile and no enum values for it, or if the enum values suggest it should. If there's enum values tagged for an SDR gamut & transfer (e.g. sRGB) *and* the file is tagged as 8-bit, then I think djxl should output 8-bit PNGs.

**Comment by dominikhlbg** [comment]
If i have an 8-bit PNG as input which i encode with jxl , then decode again but then i don't get 8-bit again because ssimulacra2 generates minus values, then something else is not working well?

**Comment by sboukortt** [comment]
If writing to an integer format, we should also not use sRGB primaries as it could cause gamut clipping (or worse). Which primaries to use is not obvious but P3 and Rec. 2020 would probably both be safer choices. (Not perfect but we don’t [need](https://en.wikipedia.org/wiki/Nirvana_fallacy#Perfect_solution_fallacy) to have found “perfect” to implement an improvement.)

Today, I started writing some logic in that direction but haven’t gotten it to work yet. What I was going for was along the lines of:

If the color space is described by an ICC profile rather than enums, and no color space hint was provided on the command line, then request the output in a color space with:

- white point: D65;
- transfer function: linear if floating point output, otherwise sRGB;
- primaries: sRGB (Rec. 709) if floating point output, otherwise P3 if 8-bit output or Rec. 2020 if 16-bit output;
- rendering intent: relative.

Any thoughts on that?

> Make cjxl recognize Display P3 and use the Enum representation for it rather than a generic icc profile. This would help for compression, and it circumvents the djxl bug.

Is that what the embedded profile is here? I’m surprised we don’t recognise it since it’s just `RGB_D65_DCI_Rel_SRG`.

**Comment by thebombzen** [comment]
> It makes sense that we currently don’t recognize it as we make no attempt to recognize pure gamma curves.

Why not? It seems very doable if there's a `curv` tag rather than a gamma LUT.

**Comment by sboukortt** [comment]
> > It makes sense that we currently don’t recognize it as we make no attempt to recognize pure gamma curves.
> 
> Why not? It seems very doable if there's a `curv` tag rather than a gamma LUT. Adobe RGB can be represented with gamma TRC and custom primaries, so I feel like we should do that.

The best answer I can give to “why not” is pretty much “because we have not implemented it so far”, but I have just started giving this a try (#2341). I have it working with skcms, now I just need to implement the lcms2 version. (Or I suppose we could already merge the skcms part, but something makes me uneasy about not having feature parity between the two.)

**Comment by sboukortt** [comment]
#2341 has now been merged. It doesn’t solve the issue with this specific `.jxl` file, but at least, if the same source file is compressed to JPEG XL again, it should create a file without the issue.

**Comment by o-l-a-v** [comment]
I'm a beginner at this, sorry if this info isn't relevant.

One more image that decodes to something that does not look like the original, but a dark and dull looking PNG.
I used `djxl v0.9.0 96d76a3` from: https://artifacts.lucaversari.it/libjxl/libjxl/2023-10-17T11%3A15%3A45Z_56949f8a9772c0139803c25e46bd8ccf8ef93f7e/

Sample image:

* Found here: https://people.csail.mit.edu/ericchan/hdr/hdr-jxl.php
* JXL: https://people.csail.mit.edu/ericchan/hdr/jxl_images/20140606_102418_IMGP0297.jxl
* Reference JPEG image / how it should look, i think: https://people.csail.mit.edu/ericchan/hdr/jxl_images/20140606_102418_IMGP0297.jpg

Reference JPEG image:

<details><summary>Click to expand</summary>

![sample_jpeg](https://people.csail.mit.edu/ericchan/hdr/jxl_images/20140606_102418_IMGP0297.jpg)

</details>

First tried `djxl in.jxl out.png --bits_per_sample=8` (with or without `--bits_per_sample=8` made no visual difference), it created a dark and dull result.

<details><summary>Click to expand</summary>

![out_default](https://github.com/libjxl/libjxl/assets/6450056/b4c22229-0a9f-4847-ac63-167882c2382a)

</details>

Then opened original JXL in GIMP to look for metadata, it said `Exif.Photo.ColorSpace: sRGB`. Then tried `djxl in.jxl out.png --color_space=RGB_D65_SRG_Per_SRG --bits_per_sample=8`, it looked good compared to the reference JPG file, and how GIMP decodes the JXL.

<details><summary>Click to expand</summary>

![out_srgb](https://github.com/0xC0000054/pdn-jpegxl/assets/6450056/0248771d-a9de-4fe9-8173-a43c51de0258)

</details>

I also tried `jxlinfo in.jxl`, it said nothing about sRGB:

```
JPEG XL file format container (ISO/IEC 18181-2)
JPEG XL image, 1638x2048, lossy, 16-bit RGB
Color space: RGB, D65, Rec.2100 primaries, PQ transfer function, rendering intent: Perceptual
Uncompressed Exif metadata: 818 bytes
Brotli-compressed XML  metadata: 5664 compressed bytes
unknown box: type: "hrgm" size: 147802
layer: full image size, name: "main"
```

**Comment by o-l-a-v** [comment]
Two images that [v0.9.1](https://github.com/libjxl/libjxl/releases/tag/v0.9.1) still have problems with.

* Jungle HDR image
  * Found here: <https://people.csail.mit.edu/ericchan/hdr/hdr-jxl.php>
  * JXL file: <https://people.csail.mit.edu/ericchan/hdr/jxl_images/20140606_102418_IMGP0297.jxl>
* Cat HDR image
  * Found here: <https://github.com/d2phap/ImageGlass/issues/1677#issuecomment-1819045271>
  * JXL file: <https://github.com/gwy15/hdr/blob/main/IMG01623-HDR.jxl>

Both `djxl input.jxl output.png` and `djxl input.jxl output.png --bits_per_sample=8` results in an output with washed out colors.

**Comment by Traneptora** [comment]
Do note that this may be an error with the actual PNG viewer. `djxl input.jxl output.png` for the Jungle image results in a PNG file that has `cICP` tags setting the color space information. If your viewer doesn't cICP, then the image will look like this:
![test-libjxl-8bit](https://github.com/libjxl/libjxl/assets/1342577/55495b0a-e42d-4bba-af86-8150afc337e7)

**Comment by sboukortt** [comment]
Even then, we used to also generate a fallback ICC chunk. That we don’t anymore is a regression, apparently introduced by #2635. I’ll have a closer look next week.

**Comment by Traneptora** [comment]
Likewise, the image on the right looks exactly like what you'd get if you interpret the BT2100/PQ png that was output as though it were sRGB.

**Comment by kmilos** [comment]
> PNG spec isn't clear what happens if you have both an `sRGB` chunk and an `iCCP` chunk. It recommends against it, but it doesn't forbid it

I think it does forbid it. But that doesn't mean it doesn't happen, or if various implementations deal with it consistently.

**Comment by sboukortt** [comment]
> IMO fallback ICC profile is not needed in the sRGB case anyway.

Right, in the sRGB case, we didn’t, only in the cICP-but-not-sRGB case.

**Comment by o-l-a-v** [comment]
> Do note that this may be an error with the actual PNG viewer. `djxl input.jxl output.png` for the Jungle image results in a PNG file that has `cICP` tags setting the color space information. If your viewer doesn't cICP, then the image will look like this:

Darktable is said to support cICP since 4.2.0: https://www.darktable.org/2022/12/darktable-4.2.0-released/

With Darktable v4.6.0, opening the PNG results in a very dark looking image:

<img width="400" alt="image" src="https://github.com/libjxl/libjxl/assets/6450056/b5289737-03e5-4060-b509-db253182734f">

I tried the latest versions of IrfanView, XnView MP, Windows Photos app, GIMP, Krita, ImageGlass, JPEGView. All view this image like the example you posted.

Do we have any example of a program that can properly view PNG with cICP?

It doesn't make sense to decode to some exotic output that no current image viewers and editors can view properly IMO.

**Comment by kmilos** [comment]
> With Darktable v4.6.0, opening the PNG results in a very dark looking image:

If this is a PQ or HLG transfer curve image, darktable does not do any tone mapping, and just pulls the image down so max stays at 1.0 (which makes an HDR image datk). Use the exposure module to push it back up.

**Comment by Traneptora** [comment]
> Do we have any example of a program that can properly view PNG with cICP?

Chromium and FFmpeg both support cICP chunks.

**Comment by jonnyawsom3** [comment]
Seems like this should be fixed, but the original JXL file has expired since Discord added time limits to their CDN.

**Comment by o-l-a-v** [comment]
@jonnyawsom3 The original image was saved by archive.org to my surprise.

* <https://web.archive.org/web/*/cdn.discordapp.com/attachments/794206170445119489/1083872278128885790/imagetest15-Q70.jxl>

I downloaded it and ZIP-ed it (just in case):

[imagetest15-Q70.zip](https://github.com/user-attachments/files/18469424/imagetest15-Q70.zip)

**Comment by jonnyawsom3** [comment]
Just found this again, seems to be fixed, but depends on the image viewer having PNG ICC support.

**Comment by sscherfke** [comment]
Not sure if this is related or not, but this issue is the closed thing to my problem that I could find.  My problem is atually the other way around.  I export HDR PNG images with Lightroom and convert them with `cjxl img.png img.jxl`.

The images look right (== HDRy) on HDR displays (Macbook, HDR TV) and they also look fine on normal/old SDR monitors connected to my Macbook. However, if viewed on Linux (checked Alma9 and Fedora 42), they are way too dark.


Version used for encoding: cjxl 0.11.1 [NEON] on macOS with Silicon/ARM.
Version used for decoding: Should also be 0.11.1 (on Fedora 42)

I can provided example images if needed.

**Comment by jonnyawsom3** [comment]
Sounds like the Linux viewers aren't color managing or using HDR, have you checked their preferences for related settings?

**Comment by sscherfke** [comment]
I attached an image with several different exports.  They even look different in the preview and it doesn't change when opened in Gwenview or Okular.

The JPG variants always look right (although, the HDR jpg displays a littble bit different than the JXLs on an HDR display).

Interestingly, the direct JXL export with maximized compatibility is not actually an HDR image but seems to be a toned mapped SDR image – it looks the same on an HDR and on an SDR display.

<img width="1302" height="999" alt="Image" src="https://github.com/user-attachments/assets/34168116-d870-436a-925a-00991ca1f6da" />

**Comment by eclipseo** [comment]
> Not sure if this is related or not, but this issue is the closed thing to my problem that I could find. My problem is atually the other way around. I export HDR PNG images with Lightroom and convert them with `cjxl img.png img.jxl`.
> 
> The images look right (== HDRy) on HDR displays (Macbook, HDR TV) and they also look fine on normal/old SDR monitors connected to my Macbook. However, if viewed on Linux (checked Alma9 and Fedora 42), they are way too dark.
> 
> Version used for encoding: cjxl 0.11.1 [NEON] on macOS with Silicon/ARM. Version used for decoding: Should also be 0.11.1 (on Fedora 42)
> 
> I can provided example images if needed.

Could you provide a sample of an image that is wrongly decoded in Fedora 42 ? With which software ? And which compositor you are using ?

**Comment by sscherfke** [comment]
Here are the two files (had to copress them with gzip, b/c GitHub does not allow jxl attachments …).

One has been been directly exported by Lightroom Classic _without_ the "maximizie compatibility" option. The other one was exported as PNG and converted with `cjxl` 0.11.1 with default options on the command line. Colorspace for both images was set to `sRGB (Rec. 709)`.

For decoding, I was using Gwenview on KDE on Fedora 42 (I guess with Wayland (the bootable live system for the Fedora KDE spin)) and on Alma 9 with X.

Please let me know if I can do anything else to help with debugging. :-)

[DSC00299-hdr-lr.jxl.gz](https://github.com/user-attachments/files/22707534/DSC00299-hdr-lr.jxl.gz)
[DSC00299-hdr-png.jxl.gz](https://github.com/user-attachments/files/22707535/DSC00299-hdr-png.jxl.gz)

**Comment by sscherfke** [comment]
And the source PNG exported via LR Classic: 
<img width="1000" height="563" alt="Image" src="https://github.com/user-attachments/assets/61ccc4aa-a770-41e8-959e-dca9facc215a" />

**Comment by jonnyawsom3** [comment]
Turns out libjxl is still defaulting to Linear, but it's hard to see in correctly color managed viewers.
It still causes excessive color errors though, which caused multiple other issues to be created since.
